File: androidApp/features/notedetail/src/main/kotlin/com/velkonost/getbetter/android/features/notedetail/components/ActionButtons.kt
package com.velkonost.getbetter.android.features.notedetail.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.velkonost.getbetter.core.compose.components.BottomActionButton
import com.velkonost.getbetter.shared.features.notedetail.presentation.contract.NoteState
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.painterResource

@Composable
fun ActionButtons(
    modifier: Modifier = Modifier,
    noteState: NoteState,
    onEditClick: () -> Unit,
    onDeleteClick: () -> Unit,
    onSaveClick: () -> Unit,
    onCancelSaveClick: () -> Unit
) {

    AnimatedVisibility(
        visible = noteState == NoteState.Editing,
        enter = fadeIn(),
        exit = fadeOut()
    ) {
        Column {
            Spacer(modifier = modifier.weight(1f))
            Row {
                CancelSaveButton(onClick = onCancelSaveClick)
                SaveButton(onClick = onSaveClick)
            }
            Spacer(modifier = modifier.height(64.dp))
        }
    }

    AnimatedVisibility(
        visible = noteState == NoteState.View,
        enter = fadeIn(),
        exit = fadeOut()
    ) {
        Column {
            Spacer(modifier = modifier.weight(1f))
            Row {
                DeleteButton(
                    onClick = onDeleteClick
                )

                EditButton(
                    onClick = onEditClick
                )
            }
        }
    }
}

@Composable
internal fun RowScope.SaveButton(onClick: () -> Unit) {
    BottomActionButton(
        icon = painterResource(imageResource = SharedR.images.ic_save),
        onClick = onClick
    )
}

@Composable
internal fun RowScope.CancelSaveButton(onClick: () -> Unit) {
    BottomActionButton(
        icon = painterResource(imageResource = SharedR.images.ic_cancel),
        onClick = onClick
    )
}

@Composable
internal fun RowScope.DeleteButton(onClick: () -> Unit) {
    BottomActionButton(
        icon = painterResource(imageResource = SharedR.images.ic_trash),
        onClick = onClick
    )
}

@Composable
internal fun RowScope.EditButton(onClick: () -> Unit) {
    BottomActionButton(
        icon = painterResource(imageResource = SharedR.images.ic_edit),
        onClick = onClick
    )
}

File: iosApp/iosApp/Features/NoteDetail/components/ActionButtons.swift
//
//  ActionButtons.swift
//  iosApp
//
//  Created by velkonost on 03.11.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import SharedSDK

struct ActionButtons: View {

    let noteState: NoteState
    let onEditClick: () -> Void
    let onDeleteClick: () -> Void
    let onSaveClick: () -> Void
    let onCancelSaveClick: () -> Void
    
    init(noteState: NoteState, onEditClick: @escaping () -> Void, onDeleteClick: @escaping () -> Void, onSaveClick: @escaping () -> Void, onCancelSaveClick: @escaping () -> Void) {
        self.noteState = noteState
        self.onEditClick = onEditClick
        self.onDeleteClick = onDeleteClick
        self.onSaveClick = onSaveClick
        self.onCancelSaveClick = onCancelSaveClick
    }

    var body: some View {
        ZStack {
            VStack(spacing: 0) {
                Spacer()
                
                HStack {
                    if noteState == NoteState.editing {
                        CancelSaveButton
                        SaveButton
                    } else {
                        DeleteButton
                        EditButton
                    }
                }
            }
            .padding(.bottom, 80)
        }
        .animation(.easeInOut(duration: 0.5), value: noteState)
    }
}

extension ActionButtons {
    private var EditButton: some View {
        BottomActionButton(
            icon: SharedR.images().ic_edit.toUIImage()!,
            onClick: onEditClick
        )
    }
    
    private var DeleteButton: some View {
        BottomActionButton(
            icon: SharedR.images().ic_trash.toUIImage()!,
            onClick: onDeleteClick
        )
    }
    
    private var SaveButton: some View {
        BottomActionButton(
            icon: SharedR.images().ic_save.toUIImage()!,
            onClick: onSaveClick
        )
    }
    
    private var CancelSaveButton: some View {
        BottomActionButton(
            icon: SharedR.images().ic_cancel.toUIImage()!,
            onClick: onCancelSaveClick
        )
    }
}
